src/MOLECULE/bond_quartic_breakable.cpp

```diff
@@ -92,42 +112,44 @@ void BondQuartic::compute(int eflag, int vflag)
 
     rsq = delx * delx + dely * dely + delz * delz;
 
-    // if bond breaks, set type to 0
-    //   both in temporary bondlist and permanent bond_type
-    // if this proc owns both atoms,
-    //   negate bond_type twice if other atom stores it
-    // if other proc owns 2nd atom, other proc will also break bond
-
+    // quartic term
     if (rsq > rc[type] * rc[type]) {
-      bondlist[n][2] = 0;
-      for (m = 0; m < atom->num_bond[i1]; m++)
-        if (atom->bond_atom[i1][m] == atom->tag[i2]) atom->bond_type[i1][m] = 0;
-      if (i2 < atom->nlocal)
-        for (m = 0; m < atom->num_bond[i2]; m++)
-          if (atom->bond_atom[i2][m] == atom->tag[i1]) atom->bond_type[i2][m] = 0;
-      continue;
+      // for breakable bonds
+      // if bond breaks, set type to 0
+      //   both in temporary bondlist and permanent bond_type
+      // if this proc owns both atoms,
+      //   negate bond_type twice if other atom stores it
+      // if other proc owns 2nd atom, other proc will also break bond
+      if (BREAKABLE) {
+        bondlist[n][2] = 0;
+        for (m = 0; m < atom->num_bond[i1]; m++)
+          if (atom->bond_atom[i1][m] == atom->tag[i2]) atom->bond_type[i1][m] = 0;
+        if (i2 < atom->nlocal)
+          for (m = 0; m < atom->num_bond[i2]; m++)
+            if (atom->bond_atom[i2][m] == atom->tag[i1]) atom->bond_type[i2][m] = 0;
+        fbond = 0.0;
+        if (EFLAG) ebond = 0.0;
+      } else {
+        // unbreakable bond
```

Comment:
This part does not make much sense to me and neither the addition of the LJ repulsion below.
Why not have the quartic potential only for the "unbreakable" bond, regardless of its length?
Having no force and constant energy does not make much sense.
And then do the switch and LJ contribution only for the breakable version.

Reply:
Oops, my mistake here. Breakable bonds should be broken here. The force and energy calculation below should be skipped. For unbreakable bonds, we still need to check with cutoff to decide if it's constant energy.

Reply:
You are still not making sense. For "unbreakable" bonds there should be no "cutoff". That is what makes them unbreakable. You seem to be confusing the meaning of R_c.  R_c is no "cutoff". Just plot the potential function for a few choices of parameters and you should see.

Reply:
Hmm, seems we are indeed talking about different things. The "breakable" bonds are broken once r > rc, and never "rebond", which means fbond=0 and ebond=0 once r>rc. I'm trying to introduce the "rebond" capability of this bond type, hence used fbond=0 and ebond=U_0 beyond rc, so that the force due to the bond is the same in two variants. And bondlist[i][2] is never set to 0. That's what "unbreakable" means to me. The potential in this case is with single minimum, and ebond = U_0 when r goes to infinity (assuming B1*B2 < 0, as in the example in the doc). 

By the way, it is producing the same forces with pair_style list quartic, where fbond=0 and ebond=0 when r>rc (differ with bond quartic by U_0 only). (Maybe you didn't intend to include it in this form?)

I guess your potential function for the unbreakable bond is the real double-well quartic function, where the bonds don't break even without a cutoff. But this form is never available in LAMMPS, to my knowledge.

Now when thinking about it, it is confusing to use a quartic functional form but only use the r<rc part. But it provides an effectively breaking bond, with smooth energy, at a specified distance with "rebond" capability, which I cannot find in other bond types. 

If it is decided that this type of bond is not desirable, I can always go back to use bond_style table, of course.

Reply:
> That's what "unbreakable" means to me

That makes no sense to me. And it is meaningless in LAMMPS because this way the bonded atoms can move an arbitrary distance, which conflicts with the domain decomposition parallelization.

Also, it should be possible to reform a bond with some atom from a *different* bond that was broken and that is not possible with this style. To model bonds that can break and then form again, you can use fix bond/react.

> By the way, it is producing the same forces with pair_style list quartic, where fbond=0 and ebond=0 when r>rc (differ with bond quartic by U_0 only). (Maybe you didn't intend to include it in this form?)

Yes, that was not intended. I will have to revise that implementation.

> Please advise how to proceed. Thanks.

Let me discuss with the other LAMMPS developers. There are things that are strange about the upstream implementation and others that make no sense to me, so I need to figure out who is using it for what and then look into how to go on. The sad part is that it now looks quite likely that we need to abandon this pull request and our combined efforts will be mostly useful toward our own better understanding of the situation.

Reply:
@akohlmey 

> That makes no sense to me. And it is meaningless in LAMMPS because this way the bonded atoms can move an arbitrary distance, which conflicts with the domain decomposition parallelization.

I realize that from the beginning, I focused on this special case of using this bond type for specific atom pairs that belong to the same molecule. In such cases, the distance between these two atoms is bounded because they still belong to the same molecule. Hence if the communication cutoff is large enough, for example larger than the size of the molecule, the simulation can proceed, though slower.

> The sad part is that it now looks quite likely that we need to abandon this pull request and our combined efforts will be mostly useful toward our own better understanding of the situation.

It's not sad for me at all :)  Looking forward to your decisions.

Reply:
@TerrySuun as expected, after my discussion with other LAMMPS developers, the consensus is that you or others with a similar interest should look into using fix bond/react. Of course you are free to keep using your modifications, but we will not integrate them into the LAMMPS distribution.

Reply:
@akohlmey 
OK, understand. But I'm not going to use fix bond/react. I don't want to change my topology during simulation. Thanks for all the help.

Reply:
@TerrySuun I suggest you talk to Jake Gissinger (author of fix bond/react) to see if what you have in mind is possible wirh fix bond/react.  It can certainly allow bonds to break and reform.  I'm not sure what you mean by "dont want to change topology".  When the bond breaks there has to be some topological change.  For example, I assume you don't want to
continue to compute 3- and 4-body interactions that include the broken bond.  What is computationally implausible about your model is that you want to check the distance between 2 broken-bond atoms no matter how far apart they are,  That's difficult to do correctly in a parallel code.  You say they won't be too far apart, but how can that be the case for any usage of the model, e.g. one which strains the system until bonds break?  Also it seems physically implausible to require the same 2 atoms rebond (requiring some memory of previous toplogy).  How would that be the case in a real material.  Why wouldn't rebonding occur when any two atoms with the correct local topology got close enough together (e.g. atoms which were previously in 2 different molecules)?  All of that is something I believe fix bond/react can do more naturally that what you are proposing, both computationally and physically.

Reply:
@sjplimp Thanks for the suggestion. My situation is that we have a model for a complex, and we're trying to "patch" it to fit better to some data. Now our model is not "flexible" enough, hence the idea of replacing certain harmonic bonds with "diminishing" bonds that provide zero force at long bond length. I agree this is a special, even weird model, even though it is quite similar to bond_style morse. (BTW, I've tried bond_style morse, which is too soft on the attractive part.) But this is probably the most straightforward way of doing it, as re-calibrate the whole model is quite time-consuming. This is indeed not a good model for general use, especially when considering the problem in parallel simulations. But I still don't see fix bond/react suitable for our need, as we'd like to keep the 1-2, 1-3 and 1-4 pair interaction constant throughout the simulation.

> I'm not sure what you mean by "dont want to change topology". When the bond breaks there has to be some topological change. For example, I assume you don't want to continue to compute 3- and 4-body interactions that include the broken bond. 

I don't want the bond to break, just to provide zero force at long distance (I know it's technically broken. Weird.). Something "bond_style harmonic/shift/cut" provides, but with continuous force at cutoff. All other interactions, 1-2, 1-3 pairs, 3-,4-body interactions, are still calculated the same way all the time. Just the bond energy can go to a specific constant, and bond force can be zero.

> What is computationally implausible about your model is that you want to check the distance between 2 broken-bond atoms no matter how far apart they are, That's difficult to do correctly in a parallel code. 

Yes, this is not good. But I don't think it is much different from bond_style morse, where you can also have very long bond lengths.

> Also it seems physically implausible to require the same 2 atoms rebond (requiring some memory of previous toplogy). 

The memory is already there, in bondlist. And hey, are harmonic bonds *physically plausible* at all times? How about bond_style harmonic/shift/cut? They can "rebond" if I understand the doc correctly.

> How would that be the case in a real material. Why wouldn't rebonding occur when any two atoms with the correct local topology got close enough together (e.g. atoms which were previously in 2 different molecules)? 

It's just a model, to hopefully give rise to two states. And it's supposed to be just a "patch". These questions are equally valid for bond_style harmonic/shift/cut. 

******
I think this PR just provides one more functional form for bonds. The raised problems based on physical and practical reasons are already there with other bond types. Maybe just let the users judge and decide, with proper cautions and warnings?

Anyway, thanks for your time and effort. I'll try to see if there's other way around. Now bond_style table is on the top of my list. Hopefully we can find some way to fit the experimental data with a small set of parameters.

Reply:
ok - thanks for the clarification - I see more clearly now what you are trying to do.  I think bond_style table could do what you want, but you will have to parameterize it yourself obviously.  If you want to do something like bond morse or bond harmonic/shift/cut, just with a different functional form, then that would be fine.  But I suggest you divorce it from the existing bond quartic (not an option, not a derived class, not the same name).  This is b/c bond quartic is associated with breakable bonds for LAMMPS users and they are conceptually different than static bonds, both in the way they are used, and internally.   And as you said you really want a regular static bond (never breaks), just one whose force goes to zero at sufficient distance.  I agree that could be documented so users understand its limitations when atoms get too far apart, and LAMMPS will throw an error in that case anyway, if a processor can't find both atoms.

I've suggested to Axel that we rename bond quartic to bond quartic/break to emphasize it is is a breakable bond.
So you could name your new style something like bond quartic/cut (similar to harmonic/shift/cut) if you want to use a quartic formula with an added cutoff feature.

Reply:
@sjplimp Thanks for the reply.
 
> I've suggested to Axel that we rename bond quartic to bond quartic/break to emphasize it is is a breakable bond. So you could name your new style something like bond quartic/cut (similar to harmonic/shift/cut) if you want to use a quartic formula with an added cutoff feature.

This makes sense. I started with quartic bond only because the functional form is flexible enough for my needs, yet with a small number of parameters.